### PR TITLE
Redesign subdir-libs and plugin scaffolding

### PR DESCRIPTION
**Pull request by Neverlord**
Followup / refinement for https://github.com/zeek/zeek/pull/2710.

The previous PR ended up touching too many things all at once. So I've pulled out the "heavy hitters" that IMHO already make for a pretty massive change-set:
- Completely re-do the way we deal with plugin BIFs. Instead of playing cache-variable ping-pong between `ZeekPluginStatic.cmake` and `src/CMakeLists.txt`, we simply write the auto-generated BIFs loader files incrementally. The `bro_bif_create_includes` and `bro_bif_create_register` functions are dead and gone.
- Provide a new `zeek_add_plugin` function that deals with all things plugin at one side, declaratively. This replaces the old begin/end functions. They still exist as to not break legacy code right away. Eventually, I'd like seeing them go as well after some deprecation cycle. Even the `bro_`-prefixed functions are still around.
- `bro_add_subdir_library` has been replaced with `zeek_add_subdir_library`. Much like the plugin function, this function is a single, declarative way to define "subdir-libraries" in Zeek now.
- Completely overhaul how dynamic plugins locate things inside of Zeek's build tree. The dynamic plugins used to guess some paths and then read the CMake cache from Zeek to "re-construct" the path and target structure. Just figuring out how this worked previously took me a while and I doubt that the people that once wrote that still remember the nitty-gritty details. So instead of doing that, Zeek writes a CMake package file now. In the build tree, there's also a special file generated from `ZeekPluginConfig.cmake.in` that we don't install. This file contains a couple paths specifically needed for building dynamic plugins. This makes it much more comprehensible how things tie together - and much easier to extend.
- New target structure. There's `zeek_internal` now. This is meant as the base target for everything else that makes sure all Zeek targets use the same compiler flags. The main targets (previously `zeek` and `libzeek`) moved to the main `CMakeLists.txt` and are now called `zeek_exe` and `zeek_lib`. There are also a few custom/interface targets to help with dependencies and passing down flags to plugins.

This PR is quite large as-is. These things are so interwoven that it's hard to separate out parts. If this is too large, I could try to push the subdir-library stuff to a followup and/or revert some of the target restructuring. I hope it's all right as-as, though. @timwoj already did a pass on the previous PR, which had most of this work except some extra setup to get plugins to find Broker.

This PR leaves all the global variables in our CMake mostly untouched. I've reverted most changes to that again and plan to do that on a followup PR (or maybe several followups).

### CODE REVIEW

File: src/CMakeLists.txt
Comment (timwoj) [inline src/CMakeLists.txt]: This whole section feels sort of obnoxious. We really have header files in all of these directories?

File: src/ZeekConfig.cmake.in
Comment (timwoj) [inline src/ZeekConfig.cmake.in:11]: Add a comment here that for MSVC, these dependencies come from Conan which is why they're spelled out explicitly.

File: src/ZeekPluginConfig.cmake.in
Comment (timwoj) [inline src/ZeekPluginConfig.cmake.in:35]: I assume this can be follow-on work?

File: src/probabilistic/CMakeLists.txt
Comment (timwoj) [inline src/probabilistic/CMakeLists.txt:12]: This is 7 lines to do 3 lines worth of work 😃

**Review by timwoj — COMMENTED** [review]
I just had some general comments/questions. Nothing significant.

File: src/ZeekPluginConfig.cmake.in
Comment (Neverlord) [reply src/ZeekPluginConfig.cmake.in:35]: Yep, I plan on doing that as a follow-up.

File: src/probabilistic/CMakeLists.txt
Comment (Neverlord) [reply src/probabilistic/CMakeLists.txt:12]: We could also add a new 'keyword' to `zeek_add_subdir_library` for doing that. Maybe `ALT_BIFS` or something like that. That would make the call side less verbose. :)

File: src/CMakeLists.txt
Comment (Neverlord) [reply src/CMakeLists.txt]: Good question. So far I've been just porting things over without trying to minimize them. Let me try if some are redundant.

File: CMakeLists.txt
Comment (Neverlord) [reply CMakeLists.txt:780]: Sure, I'm happy to rip out `BROKER_ROOT_DIR`.

File: src/probabilistic/CMakeLists.txt
Comment (timwoj) [reply src/probabilistic/CMakeLists.txt:12]: That would be better. I need to look back at what "alternative mode" is too, as it feels like there has to be a better name for that.

Comment (Neverlord) [reply src/probabilistic/CMakeLists.txt:12]: Naming is hard. 😉

Comment (timwoj) [reply src/probabilistic/CMakeLists.txt:12]: Ah, it's a thing from `bifcl`. Unfortunately, there isn't any documentation for the arguments taken by that. Bifcl.cmake refers to a `-a` argument, but then uses `-s` later. Looking at the bifcl source, there's a `-p` argument as well. Internally, both set a variable named `alternative_mode` to true. I'm going to guess we could rename it to something like "static mode", which is more descriptive about what it actually does.

**Comment by Neverlord** [comment]
@timwoj not sure what to do about the remaining CI errors:

cluster_testing:
```
[17:23:28.200] [#2] tests.persistence-state-restart-controller ... failed
[17:23:28.200]   % 'bash /zeek/testing/external/zeek-testing-cluster/.tmp/tests.persistence-state-restart-controller/persistence-state-restart-controller.sh' failed unexpectedly (exit code 100)
[17:23:28.200]   % cat .stderr
[17:23:28.200]   Creating network "tests_persistence_state_restart_controller__default" with the default driver
[17:23:28.200]   Creating tests_persistence_state_restart_controller__controller_1 ...
[17:23:28.200]   Creating tests_persistence_state_restart_controller__client_1     ...
[17:23:28.200]   Creating tests_persistence_state_restart_controller__controller_1 ... done
[17:23:28.200]   Creating tests_persistence_state_restart_controller__client_1     ... done
[17:23:28.200]   W: Failed to fetch http://deb.debian.org/debian/dists/bullseye/InRelease  Connection timed out [IP: 199.232.30.132 80]
[17:23:28.200]   W: Some index files failed to download. They have been ignored, or old ones used instead.
[17:23:28.200]   E: Package 'procps' has no installation candidate
[17:23:28.200]   E: Package 'sqlite3' has no installation candidate
```

fedora 37:
```
+--------------------------------------------------------------+
| Sat Feb 25 16:09:26 UTC 2023                                 |
| Running baseline tests: zeek                                 |
+--------------------------------------------------------------+

Timed out!
```

freebsd14:
```
Updating FreeBSD repository catalogue...
FreeBSD repository is up to date.
All repositories are up to date.
pkg: No packages available to install matching 'py39-pip' have been found in the repositories
```

I don't think these errors come from my CMake changes. 🙂

Other than that, I think we can wrap up this first round and do followups. Any preference for pulling this in? Should I rebase + squash after you've reviewed the delta since the last review?

**Comment by timwoj** [comment]
>  tests.persistence-state-restart-controller

Ignore this one. There's some timeout issues we've been fighting with one of the debian repos. I'm about to just specify a repo to use instead of letting it try to autodiscover one.

> freebsd14:

Rebase onto master. This one should be fixed.

> fedora 37

Also rebase onto master. There's some test that periodically hangs and causes the tests to timeout, but we previously didn't store enough information about what tests were passing. I made a change on master last week that should help us track it down.

File: CMakeLists.txt
Comment (timwoj) [inline CMakeLists.txt:1282]: Can this section go away then?

**Review by timwoj — COMMENTED** [review]
One extra question

File: CMakeLists.txt
Comment (Neverlord) [reply CMakeLists.txt:1282]: I've commented out installing the CMake package files for now, but I think it's still the way to go. I left this TODO as a reminder to do a followup to check if break any of our packages (thinking about OBS). I don't expect any issues from installing them, but I'd rather evaluate that separately from this big set of changes. 🙂

**Comment by Neverlord** [comment]
Still no luck with the `cluster_testing`:

```
[18:00:28.039] [#3] tests.log-archival ... failed
[18:00:28.039]   % 'bash /zeek/testing/external/zeek-testing-cluster/.tmp/tests.log-archival/log-archival.sh' failed unexpectedly (exit code 100)
[18:00:28.039]   % cat .stderr
[18:00:28.039]   Creating network "tests_log_archival__default" with the default driver
[18:00:28.039]   Creating tests_log_archival__controller_1 ...
[18:00:28.039]   Creating tests_log_archival__client_1     ...
[18:00:28.039]   Creating tests_log_archival__client_1     ... done
[18:00:28.039]   Creating tests_log_archival__controller_1 ... done
[18:00:28.039]   W: Failed to fetch http://deb.debian.org/debian/dists/bullseye/InRelease  Connection timed out [IP: 199.232.30.132 80]
[18:00:28.039]   W: Some index files failed to download. They have been ignored, or old ones used instead.
[18:00:28.039]   E: Package 'procps' has no installation candidate
[18:00:28.039] 
[18:01:02.992] 1 of 38 tests failed
```

Plus, spicy-plugin is complaining at the moment: https://github.com/zeek/spicy-plugin/pull/192.

**Comment by timwoj** [comment]
> Still no luck with the `cluster_testing`:

You need this commit from master: https://github.com/zeek/zeek/commit/097771229c8c5d2bf1498c08b5587030df6d8371

**Comment by Neverlord** [comment]
Rebasing did the trick this time around. 👍

CI is still complaining for spicy-plugin: https://github.com/zeek/spicy-plugin/pull/192/checks?check_run_id=11725451492. The thing is... I have no idea why it's not finding the headers in the CI image. I didn't change the location and I also couldn't reproduce this on a Linux machine. So before I go and reproduce the exact environment in the CI: Is it an option to do https://github.com/zeek/spicy-plugin/issues/183 now? The build that's failing is building spicy-plugin as external plugin and also with an external Spicy, we would no longer want to support that (@rsmmr, @bbannier: please correct me if I'm wrong), making tracking this down superfluous.

File: src/analyzer/protocol/krb/CMakeLists.txt
Comment (timwoj) [inline src/analyzer/protocol/krb/CMakeLists.txt:19]: I know it was like this originally, but this should probably be an absolute path just for clarity. Same below.

**Review by timwoj — CHANGES_REQUESTED** [review]

File: src/broker/CMakeLists.txt
Comment (Neverlord) [reply src/broker/CMakeLists.txt:11]: No reason, just an oversight. Fixed.

File: CMakeLists.txt
Comment (Neverlord) [reply CMakeLists.txt:145]: It's used by `zeek_target_link_libraries`. I've moved it to the main CMake file and added comments.

**Review by timwoj — CHANGES_REQUESTED** [review]
One last set of repeat stuff from before. It looks like CI is failing too on some of the plugin tests.

File: src/iosource/pcap/CMakeLists.txt
Comment (Neverlord) [reply src/iosource/pcap/CMakeLists.txt:10]: That actually breaks Zeek (i.e., btests such as `plugins.hooks` will complain). PCAP is a plugin, not a subdir-library. That's probably why I didn't want to touch the other occurrences, but for them it's fine to use the `BIFS` block since all other instances you've pointed out are actually subdir-libraries.

**Comment by Neverlord** [comment]
> It looks like CI is failing too on some of the plugin tests.

Yep, I broke them during the last rebase of the `cmake` submodule. Fixed.

But CI broke again. 😞

centos7 (???):

```
git submodule update --recursive --init
/tmp/scripts1f19620aaeee658966ade45e40cb9a9b.sh: line 5: git: command not found

Exit status: 127
```

windows:

```
-- Conan: Version found Conan version 2.0.1
-- Conan executing: C:/Program Files/Conan/conan/conan.exe install C:/zeek/ci/windows/conanfile_windows.txt --build missing --settings arch=x86_64 --settings build_type=Release --settings compiler=Visual Studio --settings compiler.version=16 --settings compiler.runtime=MT
ERROR: The default build profile 'C:\Users\ContainerAdministrator\.conan2\profiles\default' doesn't exist.
You need to create a default profile (type 'conan profile detect' command)
or specify your own profile with '--profile:build=<myprofile>'
CMake Error at cmake/conan.cmake:651 (message):
  Conan install failed='1'
Call Stack (most recent call first):
  CMakeLists.txt:70 (conan_cmake_install)


-- Configuring incomplete, errors occurred!
See also "C:/zeek/build/CMakeFiles/CMakeOutput.log".
ninja: error: loading 'build.ninja': The system cannot find the file specified.
```

The latter probably works again after pulling in #2860 via another rebase.

More importantly, what about the spicy-plugin? Can we integrate it to the main repo now, or should I invest the time figuring out what's wrong with building it outside of Zeek with external Spicy installation?

**Comment by timwoj** [comment]
> centos7 (???):

Yeah, I have no idea how this one broke, but I'm fixing it.

> More importantly, what about the spicy-plugin? Can we integrate it to the main repo now, or should I invest the time figuring out what's wrong with building it outside of Zeek with external Spicy installation?

@bbannier @rsmmr Do you have an opinion on this one?

**Comment by rsmmr** [comment]
> > More importantly, what about the spicy-plugin? Can we integrate it to the main repo now, or should I invest the time figuring out what's wrong with building it outside of Zeek with external Spicy installation?
> 
> @bbannier @rsmmr Do you have an opinion on this one?

My plan was to integrate the Spicy plugin into the main repo once the CMake overhaul has landed. I haven't really followed here, but if it's easier to do both at the same time, you could move ahead with that and do both at the same time.

**Review by timwoj — APPROVED** [review]
Other than the question about spicy-plugin the rest of this looks good to me now.

**Comment by Neverlord** [comment]
Arne's new tests are also failing. So it's probably not just the spicy-plugin. At least the new test should give me a way to reproduce this much easier now.

**Comment by Neverlord** [comment]
One build still failing, but I'm unsure whether this error has something to do with my changes: https://cirrus-ci.com/task/6657776798138368? Did we pull in something recently that could cause that? I don't see how a CMake change could change the output for `Site::local_nets`. @awelzel, @timwoj does that maybe ring a bell?

Also, still no idea what's going on with spicy-plugin CI. My understanding is, that spicy-plugin does these steps:
- Install spicy to a custom prefix
- Build Zeek with `--disable-spicy` since we want to use our own Spicy and `--include-plugins=<path-to-spicy-plugin>` to pull in the plugin despite disabling spicy (it'll find the installed spicy).
- Run the spicy tests via `make -C tests test` (after setting `PATH` correctly).

If I do these steps, I don't get the `fatal error: zeek/zeek-config.h: No such file or directory` errors I can see [on the CI result](https://cirrus-ci.com/task/6570715932000256?logs=test_build#L16). I do get errors from baseline checks failing locally, but I don't think the CMake changes could cause that. So probably just a mismatch between the spicy-plugin version and the latest Zeek?

Since this PR is already large enough as is, I think we should pull in spicy-plugin separately. Replacing the submodule with an actual directory that has the files (probably preserving the Git history?). Once that's in, I'd rebase this once more.

Or we merge this in first and choose to ignore the CI errors on the spicy-plugin. Then integrate the spicy-plugin ASAP.

Also, is there a way to check that we don't break any of the binary builds before pulling this in? Or are we just dealing with any issues that may result when they pop up?

**Comment by timwoj** [comment]
> One build still failing, but I'm unsure whether this error has something to do with my changes: https://cirrus-ci.com/task/6657776798138368? Did we pull in something recently that could cause that? I don't see how a CMake change could change the output for Site::local_nets.

It's probably related to https://github.com/zeek/zeek/pull/2771. It looks like it started failing after that was merged. I'll do a baseline update for those tests and open a PR for it.

> Also, is there a way to check that we don't break any of the binary builds before pulling this in? Or are we just dealing with any issues that may result when they pop up?

I believe there's a way to do one-off builds from branches on OBS but I don't know what buttons to push. That's definitely an @0xxon question.

**Comment by timwoj** [comment]
> It's probably related to #2771. It looks like it started failing after that was merged. I'll do a baseline update for those tests and open a PR for it.

This should be fixed on master now.

**Comment by Neverlord** [comment]
@timwoj did you see these on centosstream8 before?

```
[#3] plugins.bifs-and-scripts ... failed
  % '${DIST}/auxil/zeek-aux/plugin-support/init-plugin -u . Demo Foo' failed unexpectedly (exit code 127)
  % cat .stderr
```

I don't think I've touched that script. 😕

The failure on `zeekctl_debian11` also doesn't seem related to me. The failing test here is `command.print`: https://cirrus-ci.com/task/6185652954333184 (something with `worker-2   Site::local_nets`).

**Comment by timwoj** [comment]
> I don't think I've touched that script.

We had a separate problem with that build. You need https://github.com/zeek/zeek/commit/79ea84e5f to fix it.

**Comment by timwoj** [comment]
Should we consider fixing this error when trying to build packages that don't specify a new enough version of CMake?

```
CMake Warning (dev) at /Users/tim/Desktop/projects/zeek-master/cmake/ZeekPlugin.cmake:53 (if):
  Policy CMP0057 is not set: Support new IN_LIST if() operator.  Run "cmake
  --help-policy CMP0057" for policy details.  Use the cmake_policy command to
  set the policy and suppress this warning.

  IN_LIST will be interpreted as an operator when the policy is set to NEW.
  Since the policy is not set the OLD behavior will be used.
Call Stack (most recent call first):
  /Users/tim/Desktop/projects/zeek-master/cmake/ZeekPluginDynamic.cmake:66 (zeek_next_pac_block)
  /Users/tim/Desktop/projects/zeek-master/cmake/ZeekPlugin.cmake:116 (zeek_add_dynamic_plugin)
  /Users/tim/Desktop/projects/zeek-master/cmake/ZeekPluginCommon.cmake:70 (zeek_add_plugin)
  CMakeLists.txt:13 (zeek_plugin_end)
This warning is for project developers.  Use -Wno-dev to suppress it.

CMake Error at /Users/tim/Desktop/projects/zeek-master/cmake/ZeekPlugin.cmake:53 (if):
  if given arguments:

    "arg" "IN_LIST" "separators"

  Unknown arguments specified
Call Stack (most recent call first):
  /Users/tim/Desktop/projects/zeek-master/cmake/ZeekPluginDynamic.cmake:66 (zeek_next_pac_block)
  /Users/tim/Desktop/projects/zeek-master/cmake/ZeekPlugin.cmake:116 (zeek_add_dynamic_plugin)
  /Users/tim/Desktop/projects/zeek-master/cmake/ZeekPluginCommon.cmake:70 (zeek_add_plugin)
  CMakeLists.txt:13 (zeek_plugin_end)
```

That's on package that requires CMake 3.0. If I bump it up to 3.15 in the package it builds just fine. Zeek itself requires 3.15, but I don't think we've ever enforced a requirement that external packages match what Zeek wants. @ckreibich What do you think here?

**Comment by Neverlord** [comment]
> Should we consider fixing this error when trying to build packages that don't specify a new enough version of CMake?

Just my 2 cents: I'd say no. You need to have a CMake >= 3.15 installed anyways for building Zeek in the first place. If we allow plugins to sit on legacy CMake versions, we make it more difficult for ourselves than we have to. It's a trivial change for plugin maintainers but non-trivial for us if we try to keep supporting legacy CMake mode; plus, we'd have to discuss how far back we wish to support. 3.0? 2.8? Any choice would be arbitrary. What I would suggest instead: have check in the plugins.cmake file whether the configured CMake version is at least what Zeek is using and otherwise raise an error to tell maintainers what to do.

**Comment by timwoj** [comment]
> What I would suggest instead: have check in the plugins.cmake file whether the configured CMake version is at least what Zeek is using and otherwise raise an error to tell maintainers what to do.

This sounds like the best way forward. Can you get that into this PR?

**Comment by rsmmr** [comment]
> This sounds like the best way forward. Can you get that into this PR?

Agree as well. Requiring the same (at least) version as Zeek itself sounds like a reasonable policy; hard to thing of else anything else that would be better.

**Comment by timwoj** [comment]
Beyond that one more change, some notes in NEWS would be great.

**Comment by timwoj** [comment]
> This sounds like the best way forward. Can you get that into this PR?

@Neverlord I pushed a commit into your cmake branch that adds this warning. I made it a WARNING, but I could see making it a FATAL_ERROR as well. When you do the rebase, make sure to pull that in too.

**Pull request merged**
